### ISSUE TITLE
Razer Mamba (1532:0024/0025) support

### ISSUE THREAD

**Issue by ghost** [openrazer/openrazer#81]
Hi! Can you add support for wireless Razer Mamba. 

lsusb -d 1532:0025 -v

Bus 001 Device 002: ID 1532:0025 Razer USA, Ltd 
Device Descriptor:
  bLength                18
  bDescriptorType         1
  bcdUSB               2.00
  bDeviceClass            0 (Defined at Interface level)
  bDeviceSubClass         0 
  bDeviceProtocol         0 
  bMaxPacketSize0         8
  idVendor           0x1532 Razer USA, Ltd
  idProduct          0x0025 
  bcdDevice            2.00
  iManufacturer           1 Razer
  iProduct                2 Razer Mamba Charging dock
  iSerial                 0 
  bNumConfigurations      1
  Configuration Descriptor:
    bLength                 9
    bDescriptorType         2
    wTotalLength           59
    bNumInterfaces          2
    bConfigurationValue     1
    iConfiguration          0 
    bmAttributes         0xa0
      (Bus Powered)
      Remote Wakeup
    MaxPower              100mA
    Interface Descriptor:
      bLength                 9
      bDescriptorType         4
      bInterfaceNumber        0
      bAlternateSetting       0
      bNumEndpoints           1
      bInterfaceClass         3 Human Interface Device
      bInterfaceSubClass      1 Boot Interface Subclass
      bInterfaceProtocol      2 Mouse
      iInterface              0 
        HID Device Descriptor:
          bLength                 9
          bDescriptorType        33
          bcdHID               1.11
          bCountryCode            0 Not supported
          bNumDescriptors         1
          bDescriptorType        34 Report
          wDescriptorLength      94
         Report Descriptors: 
           ** UNAVAILABLE **
      Endpoint Descriptor:
        bLength                 7
        bDescriptorType         5
        bEndpointAddress     0x81  EP 1 IN
        bmAttributes            3
          Transfer Type            Interrupt
          Synch Type               None
          Usage Type               Data
        wMaxPacketSize     0x0008  1x 8 bytes
        bInterval               1
    Interface Descriptor:
      bLength                 9
      bDescriptorType         4
      bInterfaceNumber        1
      bAlternateSetting       0
      bNumEndpoints           1
      bInterfaceClass         3 Human Interface Device
      bInterfaceSubClass      1 Boot Interface Subclass
      bInterfaceProtocol      1 Keyboard
      iInterface              0 
        HID Device Descriptor:
          bLength                 9
          bDescriptorType        33
          bcdHID               1.11
          bCountryCode            0 Not supported
          bNumDescriptors         1
          bDescriptorType        34 Report
          wDescriptorLength     103
         Report Descriptors: 
           ** UNAVAILABLE **
      Endpoint Descriptor:
        bLength                 7
        bDescriptorType         5
        bEndpointAddress     0x82  EP 2 IN
        bmAttributes            3
          Transfer Type            Interrupt
          Synch Type               None
          Usage Type               Data
        wMaxPacketSize     0x0009  1x 9 bytes
        bInterval               8
Device Status:     0x0000
  (Bus Powered)

lsusb -d 1532:0024 -v                                                                                                                                    
                                                                                                                                                                                
Bus 001 Device 005: ID 1532:0024 Razer USA, Ltd Razer Mamba                                                                                                                     
Device Descriptor:                                                                                                                                                              
  bLength                18                                                                                                                                                              
  bDescriptorType         1                                                                                                                                                              
  bcdUSB               2.00                                                                                                                                                              
  bDeviceClass            0 (Defined at Interface level)                                                                                                                                      
  bDeviceSubClass         0                                                                                                                                                                    
  bDeviceProtocol         0                                                                                                                                                                      
  bMaxPacketSize0         8                                                                                                                                                                           
  idVendor           0x1532 Razer USA, Ltd                                                                                                                                                            
  idProduct          0x0024 Razer Mamba                                                                                                                                                               
  bcdDevice            2.00                                                                                                                                                                                
  iManufacturer           1 Razer
  iProduct                2 Razer Mamba
  iSerial                 0 
  bNumConfigurations      1
  Configuration Descriptor:
    bLength                 9
    bDescriptorType         2
    wTotalLength           59
    bNumInterfaces          2
    bConfigurationValue     1
    iConfiguration          0 
    bmAttributes         0xa0
      (Bus Powered)
      Remote Wakeup
    MaxPower              100mA
    Interface Descriptor:
      bLength                 9
      bDescriptorType         4
      bInterfaceNumber        0
      bAlternateSetting       0
      bNumEndpoints           1
      bInterfaceClass         3 Human Interface Device
      bInterfaceSubClass      1 Boot Interface Subclass
      bInterfaceProtocol      2 Mouse
      iInterface              0 
        HID Device Descriptor:
          bLength                 9
          bDescriptorType        33
          bcdHID               1.11
          bCountryCode            0 Not supported
          bNumDescriptors         1
          bDescriptorType        34 Report
          wDescriptorLength      94
         Report Descriptors: 
           ** UNAVAILABLE **
      Endpoint Descriptor:
        bLength                 7
        bDescriptorType         5
        bEndpointAddress     0x81  EP 1 IN
        bmAttributes            3
          Transfer Type            Interrupt
          Synch Type               None
          Usage Type               Data
        wMaxPacketSize     0x0008  1x 8 bytes
        bInterval               1
    Interface Descriptor:
      bLength                 9
      bDescriptorType         4
      bInterfaceNumber        1
      bAlternateSetting       0
      bNumEndpoints           1
      bInterfaceClass         3 Human Interface Device
      bInterfaceSubClass      1 Boot Interface Subclass
      bInterfaceProtocol      1 Keyboard
      iInterface              0 
        HID Device Descriptor:
          bLength                 9
          bDescriptorType        33
          bcdHID               1.11
          bCountryCode            0 Not supported
          bNumDescriptors         1
          bDescriptorType        34 Report
          wDescriptorLength     103
         Report Descriptors: 
           ** UNAVAILABLE **
      Endpoint Descriptor:
        bLength                 7
        bDescriptorType         5
        bEndpointAddress     0x82  EP 2 IN
        bmAttributes            3
          Transfer Type            Interrupt
          Synch Type               None
          Usage Type               Data
        wMaxPacketSize     0x0009  1x 9 bytes
        bInterval               1
Device Status:     0x0000
  (Bus Powered)

ls -l /dev/input/by-id
total 0
lrwxrwxrwx 1 root root  9 nov   18 21:51 usb-Razer_Razer_Mamba_Charging_dock-event-mouse -> ../event7
lrwxrwxrwx 1 root root  9 nov   18 21:51 usb-Razer_Razer_Mamba_Charging_dock-if01-event-kbd -> ../event8
lrwxrwxrwx 1 root root  9 nov   18 21:51 usb-Razer_Razer_Mamba_Charging_dock-mouse -> ../mouse0

ls -l /dev/input/by-id
total 0
lrwxrwxrwx 1 root root  9 nov   18 21:54 usb-Razer_Razer_Mamba-event-mouse -> ../event7
lrwxrwxrwx 1 root root  9 nov   18 21:54 usb-Razer_Razer_Mamba-if01-event-kbd -> ../event8
lrwxrwxrwx 1 root root  9 nov   18 21:54 usb-Razer_Razer_Mamba-mouse -> ../mouse0

**Comment by terricain** [comment]
> What I normally do is go through the Razer Synapse app and take screenshots of all the various features and buttons you can click on.

Get me that then I can make a start.

**Comment by ghost** [comment]
Here are the screenshots. Thank you!

![mamba1](https://cloud.githubusercontent.com/assets/23561506/20474376/cd573332-afce-11e6-87c9-619cb9ac0fc1.png)
![mamba2](https://cloud.githubusercontent.com/assets/23561506/20474377/cd6f756e-afce-11e6-991d-fe12e090f886.png)
![mamba3](https://cloud.githubusercontent.com/assets/23561506/20474378/cd84c1bc-afce-11e6-8d81-ff31157b48f5.png)
![mamba4](https://cloud.githubusercontent.com/assets/23561506/20474379/cd855e1a-afce-11e6-96f5-2c413eef6ead.png)
![mamba5](https://cloud.githubusercontent.com/assets/23561506/20474380/cd861f26-afce-11e6-8c95-0c4e5816c473.png)
![mamba6](https://cloud.githubusercontent.com/assets/23561506/20474381/cd8642a8-afce-11e6-9f79-31a6e19ba6e9.png)

**Comment by terricain** [comment]
Hey, you still after this / able to help work on support?

**Comment by ghost** [comment]
I'm still after this, how can I help?

**Comment by terricain** [comment]
Need some packet captures, you capable of doing [this](https://github.com/terrycain/razer-drivers/wiki/Reverse-Engineering-USB-Protocol)?

**Comment by brwyatt** [comment]
So this is weird. The drivers seem to claim support for the Mamba (wired and wireless), but with IDs 1532:0044 and 1532:0045. However, mine (as well as nator666's) shows up as 1532:0024 and 1532:0025.

I'm guessing this might be because the DEBUG logging seems to indicate support for Mamba Chroma, so it looks like the regular Mamba (non-Chroma) isn't getting picked up.

I might be able to assist with USB debugging, though I really only have a demo install of Windows 10 available (I haven't had a valid Windows license in a long time).

**Comment by terricain** [comment]
@brwyatt Demo install "should" work fine as long as razer synapse works, is it a vm?

And yes your right the chroma one works, it didnt have chroma in the name as that was what it was called on the razer site at the time

**Comment by brwyatt** [comment]
@terrycain It is a VM. I only have Ubuntu hosts, but do have a Windows 10 (demo) VM. I can take a look at this this weekend, if I can find time.

**Comment by terricain** [comment]
Yeah it should be fine. And no rush mate

**Comment by Raboebie** [comment]
Hi Everyone. I'm facing the exact same problem. PID is listed as 0024. It's a Razor Mamba of that I'm sure :). Any update on this issue?

**Comment by terricain** [comment]
Waiting on packet captures I think

**Comment by brwyatt** [comment]
I managed to get time to do this today. Did everything (I hope/think) under lighting and power (that was configurable), and the basics under performance (enable/disable XY sensitivity, sensitivity, and polling rate; Acceleration doesn't seem to be sent to the device, and I did not do independent xy sensitivities or sensitivity stages). All of this was done with the mouse wired, but let me know if you also need these for the mouse while wireless. (I assume it should be the same, just with the different device ID)

Captures are here: http://home.brwyatt.net/files/RazerMamba2012.zip

I'm hoping this is what you needed, but let me know if there's something not right about them. I can also go ahead and get the separate XY sensitivities and the sensitivity stages as well. Additional note: I didn't capture all 63 different values for the sensitivities, but I'm hoping/assuming they can be extrapolated from the provided data. If not, let me know and I can get the remaining ones for you.

**Comment by terricain** [comment]
Looks good, could of done 1 capture per effect, you could of combined all of the colour ones into one capture :wink:, but the beauty of pcapng is i can just `cat` them all together

I dont suppose I can get a packet capture of the entire startup stopping when synapse is opened? In theory synapse should query the device for a serial.

Also on wireless could I get a capture of like 10 mins of no movement/no clicks, synapse should query the mouse for its battery status and if its plugged in charging or not.

**Comment by terricain** [comment]
Am pretty sure I know what the serial command is but having captures is nice for my records at least in case i ever need to revisit stuff. When I get home should be able to get to work and have it mostly functional. Assuming you are willing to test, what distro are you on?

**Comment by Raboebie** [comment]
I'm also available to test if you require more than one participant. I'm running Ubuntu Gnome 16.04 x64.

**Comment by terricain** [comment]
Sure, I'll basically have a list of things to get through so the more people the quicker it gets done

**Comment by Raboebie** [comment]
Excellent.

**Comment by terricain** [comment]
Can you change dock colour?

**Comment by Raboebie** [comment]
If you mean from within Synapse then no, you can't change the dock colour. It's random

**Comment by terricain** [comment]
Ahh sad times

**Comment by terricain** [comment]
@Raboebie Install debs from here - https://drive.google.com/drive/folders/0BydYgD2LzDUCdE1TSENkY0NGSzg?usp=sharing

We'll test with mamba on wireless first
Then reboot, then 
```
ls -lah /sys/bus/hid/drivers/
ls -lah /sys/bus/hid/drivers/razermouse/
ls -lah /sys/bus/hid/drivers/razermouse/0*/
```

**Comment by Raboebie** [comment]
Thank you! I'll give it a bash tomorrow and let you know.

**Comment by Raboebie** [comment]
Installed the debs. Attached are the outputs from the commands you asked for.




[commands.zip](https://github.com/terrycain/razer-drivers/files/790531/commands.zip)

**Comment by Raboebie** [comment]
For interests sake, I have been using xinput to disable to acceleration of the mounse and change the sensitivity below normal. Is that possible to do with this driver? (If you mouse is supported)

**Comment by z3ntu** [comment]
Acceleration should be something with Xorg but dpi is possible with most (if not all) mice

**Comment by terricain** [comment]
Mouse supports DPI so yep (acceleration is a software thing). We'll test some more stuff, then test the daemon, then should be done. The front end apps "should" allow you to change dpi (if they have added the option)

**Comment by Raboebie** [comment]
Oh yes I know that I was just wondering if there was an "easy" way to disable it with the daemon. I was hoping for that amazing feeling after you install synapse and the acceleration is all gone... ;). But with xorg there's never really an easy way to do anything without messing up something else in my experience. Thank you for your efforts on this. It's much appreciated!

**Comment by terricain** [comment]
No problem. Thanks for helping test. I've never seen accleration on any mice I have. What xinput stuff are you running?

**Comment by Raboebie** [comment]
So what I'm doing currently is using xinput to identify the device and list-props to get the properties. I then use set-prop like every ubuntu thread about mice would suggest but since I updated to 16.04 it doesn't work anymore. Editing the xorg.conf file directly and restarting X also doesn't work. So I'm stuck with a mouse that jumps across the screen when I breathe too heavily :P. Interestingly enough if I use the Mamba in wireless more I can't use xinput to list any of iets properties. It only lists the dock but it lists the dock twice. Since it sees it as the dock now Acceleration and Deceleration aren't "valid" properties anymore. If in wired mode I can see the properties but changing them has no effect.

**Comment by terricain** [comment]
To be fair I think just lowering the dpi should solve that or at least make it more useable ;-)

**Comment by Raboebie** [comment]
Lets hope so :)

**Comment by Raboebie** [comment]
The good news is I managed to change the DPI of my mouse using the scripts in in the project. Finally, I have a useable desktop.

**Comment by terricain** [comment]
@Raboebie :-)

**Comment by terricain** [comment]
sudo to root
cd to `/sys/bus/hid/drivers/razermouse/0003:1532:0025.0005/` or which ever one contains `dpi`

```
cat charge_level
cat charge_status
cat device_mode
cat device_serial
cat device_type
cat dpi
cat firmware_version
cat poll_Rate
cat scroll_led_brightness
cat scroll_led_effect
cat scroll_led_rgb
cat scroll_led_state
cat version
```

**Comment by Raboebie** [comment]
I'll run them for you as soon as I get home.

**Comment by Raboebie** [comment]
See attached.
[outputs.zip](https://github.com/terrycain/razer-drivers/files/793798/outputs.zip)

**Comment by terricain** [comment]
@Raboebie Would be easier if you just pasted the output into a code block

**Comment by Raboebie** [comment]
Sorry, didn't want to spam up the thread.

**Comment by terricain** [comment]
@Raboebie Ahh thats fine

**Comment by terricain** [comment]
Did cat device_serial not work?

```
# Set brightness of scroll wheel to max
echo -n '255' > scroll_led_brightness
# Half
echo -n '128' > scroll_led_brightness
# Quarter brightness
echo -n '64' > scroll_led_brightness
# Max
echo -n '255' > scroll_led_brightness

#Hopefully red
echo -ne '\xFF\x00\x00' > scroll_led_rgb
echo -ne '0' > scroll_led_effect

# Hopefully specturm
echo -ne '4' > scroll_led_effect
```

**Comment by Raboebie** [comment]
Everything that came back blank was blank when I ran it. I'll try again
tomorrow to make sure. I'll run these tomorrow morning and let you know.

**Comment by terricain** [comment]
@Raboebie 👍cool

**Comment by Raboebie** [comment]
All of those commands worked as expected, except for the last one. The last one disabled the lighting completely and on the second run did nothing. Something that's pretty cool is that when the device is in wireless more the dock actually changes colour along with the mouse when I run these commands. And yes, cat device_serial does not work.

**Comment by tildearrow** [comment]
I have a question. Why "scroll_led_" and not "matrix_effect_"? The Orochi, having 1 LED like this Mamba, uses " matrix_effect_" files.

**Comment by z3ntu** [comment]
It's a matter of how the firmware of the device is made afaik

**Comment by terricain** [comment]
@z3ntu Yep exactly. It depends on how things are done in firmware

**Comment by brwyatt** [comment]
Just getting a chance to look back at this. Do you still need further captures, @terrycain? It doesn't look like anyone else was able to provide any, but wanted to make sure. Sounds like you had enough to at least get together at least a small test driver.

If I'm not mistaken, the things you are looking for still are:

1. Captures of the startup of synapse
2. 10 minutes of no movement on wireless
3. 10 minutes of no movement while wired (It kinda sounded like this could be useful, too, was hard to tell in context for this one)

**Comment by terricain** [comment]
@brwyatt Yh

**Comment by WhyNotHugo** [comment]
Hi!

When I move my mamba's wheel to the side, the event is sends it the same as scrolling the wheel multiple times (checked this closely with `xev`).

Would adding a possible fix for the be part of this issue, or is that out of scope here? If it's part of this issue: how can I help?

I'll go ahead and start fetching a windows ISO so I can set up a VM, since I'm guessing I'll probably need data from there, right?

**Comment by terricain** [comment]
@hobarrera there is code to fix that but its waiting on other stuff but its on my list don't worry :)

**Comment by WhyNotHugo** [comment]
Thanks! Do count on my if your need any debug/test data.

**Comment by terricain** [comment]
Sure, can you get telegram and add me `@terrycain`, most of the developers hang out there and I'll have some macro based stuff to test (which will include this bit) soon

**Comment by brwyatt** [comment]
Hey guys, apologies for dropping the ball on this one. I'm going to try and get the additional stuff that was requested done today after I get home from work.

**Comment by terricain** [comment]
@brwyatt Mate no rush. Though if your around tomorrow to test could get it in the Sunday release

**Comment by brwyatt** [comment]
Synapse startup, 10 minutes wireless with no movement, 10 minutes wired with no movement (little bit of something 40 seconds in, but 10 minutes following that)

[Razer Mamba 2012 - 2.zip](https://github.com/terrycain/razer-drivers/files/923226/Razer.Mamba.2012.-.2.zip)

Hopefully that works for what you need! Let me know if you need anything further!

**Comment by terricain** [comment]
Ok anyone, get debs from here - https://drive.google.com/open?id=0BydYgD2LzDUCdE1TSENkY0NGSzg

Install them, reboot then either respond here, or on telegram if your on there.

**Comment by WhyNotHugo** [comment]
Debs don't really work for me, but I can easily work from the source branch if you point me to it. :)

**Comment by z3ntu** [comment]
@hobarrera What distro are you using? I can make you easy instructions for RPM and Arch Linux.

**Comment by WhyNotHugo** [comment]
I'm using ArchLinux. I don't think I need further instructions (since I can just tweak existing PKGBUILDs), I mostly need a confirmation of which branch this is based of and I can probably work from there.

**Comment by z3ntu** [comment]
Great, use https://aur.archlinux.org/pkgbase/razer-drivers-git/ as a base and add a `#branch=branchname` to the end of the source field (eg `source=("$pkgbase::git+https://github.com/terrycain/razer_drivers.git#branch=bug_lighting_0x0220")`)
I don't think @terrycain put the branch online yet but hopefully will soon.

**Comment by z3ntu** [comment]
Branch `feature_mamba2012_0024_0025` is online.

**Comment by WhyNotHugo** [comment]
Oh, apologies, mine's a 0045, so I don't think the changes so far apply to me. 😅 

```
$ lsusb | grep Razer
Bus 008 Device 006: ID 1532:0045 Razer USA, Ltd
```

**Comment by z3ntu** [comment]
`0045` is already supported according to the readme.

**Comment by WhyNotHugo** [comment]
Yes, sorry, I got confused. I initially jumped in because I wanted to see if I could help on *additional* features for the mamba ([like a few buttons working a bit off](https://github.com/terrycain/razer-drivers/issues/81#issuecomment-294093238)), but didn't realize the changes so far aren't related to that. 😅

**Comment by lesderid** [comment]
I tried the branch on a 0024 with polychromatic, but it doesn't seem to be doing much. Polychromatic says no devices found. Output from the daemon (started daemon, unplugged, plugged back in, opened polychromatic):
```
2017-04-15 18:16:42 | razer                          | INFO     | Initialising Daemon (v1.1.9). Pid: 15425
2017-04-15 18:16:42 | razer.screensaver              | INFO     | Initialising DBus Screensaver Monitor
(...)
2017-04-15 18:16:42 | razer                          | DEBUG    | Loaded device specification: RazerMamba2012Wired --------------------- (1532:0024)
2017-04-15 18:16:42 | razer                          | DEBUG    | Loaded device specification: RazerMamba2012Wireless ------------------ (1532:0025)
(...)
2017-04-15 18:16:42 | razer                          | INFO     | Serving DBus
2017-04-15 18:16:51 | razer                          | DEBUG    | Device event [remove]: /devices/pci0000:00/0000:00:09.0/0000:02:00.0/usb2/2-3/2-3:1.0/0003:1532:0024.0018
2017-04-15 18:16:51 | razer                          | DEBUG    | Device event [remove]: /devices/pci0000:00/0000:00:09.0/0000:02:00.0/usb2/2-3/2-3:1.1/0003:1532:0024.0019
2017-04-15 18:16:55 | razer                          | DEBUG    | Device event [add]: /devices/pci0000:00/0000:00:09.0/0000:02:00.0/usb2/2-3/2-3:1.0/0003:1532:0024.001A
2017-04-15 18:16:55 | razer                          | INFO     | Found valid device.0: 0003:1532:0024.001A
2017-04-15 18:16:55 | razer.device0                  | INFO     | Initialising device.0 RazerMamba2012Wired
2017-04-15 18:16:59 | razer                          | DEBUG    | DBus called get_serial_list
```

**Comment by terricain** [comment]
reboot, then have a look?

**Comment by lesderid** [comment]
I've already reloaded the module, but I'll try rebooting anyway.

**Comment by lesderid** [comment]
Polychromatic now times out ("Polychromatic took too long to initialise the application or daemon.") and I need to send a SIGKILL to the daemon to shut it down.

**Comment by terricain** [comment]
kill, the daemon, start it then send full log

**Comment by lesderid** [comment]
I'd already rebooted again.

Trying again now, Polychromatic doesn't timeout anymore, but still says it can't find any devices. Daemon log doesn't have anything interesting (same as before). dmesg shows this however:
```
[   63.842876] razermouse 0003:1532:0024.0005: input,hidraw4: USB HID v1.11 Keyboard [Razer Razer Mamba] on usb-0000:02:00.0-4/input1
[   63.904264] BUG: unable to handle kernel NULL pointer dereference at           (null)
[   63.904288] IP: razer_send_control_msg+0x38/0xd0 [razermouse]
[   63.904292] PGD 2ffdbf067 
[   63.904294] PUD 2ffdb7067 
[   63.904297] PMD 0 

[   63.904307] Oops: 0000 [#1] PREEMPT SMP
[   63.904311] Modules linked in: razermouse(O) razercore(O) tun fuse ebtable_filter ebtables ip6table_filter ip6_tables iptable_filter nct6775 zram hwmon_vid ipt_MASQUERADE nf_nat_masquerade_ipv4 iptable_nat nf_conntrack_ipv4 nf_defrag_ipv4 nf_nat_ipv4 nf_nat nf_conntrack libcrc32c crc32c_generic cfg80211 rfkill snd_hda_codec_hdmi snd_usb_audio input_leds led_class joydev snd_usbmidi_lib snd_rawmidi snd_seq_device evdev mac_hid nvidia_drm(PO) nvidia_modeset(PO) nvidia(PO) edac_mce_amd edac_core kvm_amd kvm irqbypass snd_hda_codec_realtek drm_kms_helper crct10dif_pclmul crc32_pclmul ghash_clmulni_intel snd_hda_codec_generic pcbc aesni_intel drm aes_x86_64 syscopyarea sysfillrect r8169 crypto_simd sysimgblt snd_hda_intel glue_helper fb_sys_fops snd_hda_codec cryptd snd_hda_core mii sp5100_tco pcspkr
[   63.904378]  k10temp fam15h_power shpchp i2c_piix4 snd_hwdep fjes acpi_cpufreq tpm_tis nuvoton_cir rc_core tpm_tis_core button tpm sch_fq_codel vboxnetflt(O) vboxnetadp(O) pci_stub vboxpci(O) vboxdrv(O) usbip_host usbip_core nfsd auth_rpcgss loop oid_registry nfs_acl dm_crypt lockd dm_mod snd_aloop grace sunrpc snd_pcm snd_timer snd soundcore ip_tables x_tables btrfs xor raid6_pq sd_mod ata_generic pata_acpi hid_generic usbhid hid ohci_pci crc32c_intel ahci xhci_pci pata_atiixp libahci xhci_hcd ehci_pci ohci_hcd ehci_hcd libata usbcore scsi_mod usb_common serio
[   63.904441] CPU: 0 PID: 3199 Comm: razer-service Tainted: P           O    4.10.1-1-ARCH #1
[   63.904446] Hardware name: To Be Filled By O.E.M. To Be Filled By O.E.M./970 Pro3 R2.0, BIOS P1.80 10/24/2014
[   63.904450] task: ffff880362e90e40 task.stack: ffffc900072ac000
[   63.904459] RIP: 0010:razer_send_control_msg+0x38/0xd0 [razermouse]
[   63.904463] RSP: 0018:ffffc900072afb08 EFLAGS: 00010286
[   63.904468] RAX: ffff8802f03266c0 RBX: 0000000000000000 RCX: ffff8802f03266c0
[   63.904471] RDX: 0000000000000002 RSI: ffffc900072afcae RDI: ffff8802f032671a
[   63.904475] RBP: ffffc900072afb48 R08: ffff8803a10037c0 R09: ffff8802f03266c0
[   63.904478] R10: 00007fe8994269e0 R11: ffff880362e90e40 R12: 0000000000000258
[   63.904482] R13: 0000000000000320 R14: ffffc900072afbc6 R15: 0000000000000000
[   63.904487] FS:  00007fe88660d700(0000) GS:ffff8803afc00000(0000) knlGS:0000000000000000
[   63.904491] CS:  0010 DS: 0000 ES: 0000 CR0: 0000000080050033
[   63.904494] CR2: 0000000000000000 CR3: 00000002ffdbd000 CR4: 00000000000406f0
[   63.904497] Call Trace:
[   63.904506]  ? kmem_cache_alloc_trace+0x19c/0x1b0
[   63.904514]  razer_get_usb_response+0x61/0x150 [razermouse]
[   63.904522]  razer_send_payload+0x58/0x1a0 [razermouse]
[   63.904530]  razer_attr_read_get_serial+0x91/0xf0 [razermouse]
[   63.904536]  ? mntput+0x24/0x40
[   63.904541]  ? terminate_walk+0xe7/0x100
[   63.904548]  dev_attr_show+0x20/0x50
[   63.904554]  sysfs_kf_seq_show+0xb9/0x130
[   63.904558]  kernfs_seq_show+0x20/0x30
[   63.904563]  seq_read+0x113/0x3d0
[   63.904568]  kernfs_fop_read+0x113/0x180
[   63.904573]  __vfs_read+0x37/0x130
[   63.904579]  ? security_file_permission+0x9b/0xb0
[   63.904583]  vfs_read+0x96/0x130
[   63.904588]  SyS_read+0x55/0xc0
[   63.904594]  entry_SYSCALL_64_fastpath+0x1a/0xa9
[   63.904598] RIP: 0033:0x7fe89981ef7d
[   63.904601] RSP: 002b:00007fe88660b3e0 EFLAGS: 00000293 ORIG_RAX: 0000000000000000
[   63.904606] RAX: ffffffffffffffda RBX: 0000000000000063 RCX: 00007fe89981ef7d
[   63.904609] RDX: 0000000000001001 RSI: 00007fe880003a60 RDI: 000000000000000b
[   63.904612] RBP: 0000000000002710 R08: 0000000000000000 R09: 00007fe880003a40
[   63.904616] R10: 00007fe8994269e0 R11: 0000000000000293 R12: 00007fe880000078
[   63.904619] R13: 0000000000001030 R14: 00007fe880003a30 R15: 00007fe880000020
[   63.904623] Code: 41 55 41 54 49 89 ff 53 48 89 f7 89 d3 be 5a 00 00 00 ba c0 00 40 01 49 89 cc 48 83 ec 18 4d 89 c5 e8 4d 1a 84 e0 48 85 c0 74 6c <41> 8b 37 44 0f b7 cb 41 b8 00 03 00 00 b9 21 00 00 00 ba 09 00 
[   63.904691] RIP: razer_send_control_msg+0x38/0xd0 [razermouse] RSP: ffffc900072afb08
[   63.904694] CR2: 0000000000000000
[   63.904698] ---[ end trace 39a3d121aa79e9f4 ]---
[   63.906406] mousedev: PS/2 mouse device common for all mice
```

**Comment by terricain** [comment]
ooh :( Will go have a look

**Comment by terricain** [comment]
Ok think it was because i forgot to call an init function. Heres new debs - https://drive.google.com/open?id=0BydYgD2LzDUCdE1TSENkY0NGSzg

Probs will need to reboot if driver doesnt remove as sometimes it refuses onces its died.

**Comment by lesderid** [comment]
Could you push the branch again? I was compiling from source on Arch Linux.

**Comment by terricain** [comment]
There you go

**Comment by lesderid** [comment]
I've added a `razer_mouse_init` call in my local copy for now (edit: same line as in your last commit). That does seem to fix the driver part.

New issue when running polychromatic:
```
ERROR: Exception: org.freedesktop.DBus.Error.UnknownMethod: Traceback (most recent call last):
  File "/usr/lib/python3.6/site-packages/dbus/service.py", line 654, in _message_cb
    (candidate_method, parent_method) = _method_lookup(self, method_name, interface_name)
  File "/usr/lib/python3.6/site-packages/dbus/service.py", line 246, in _method_lookup
    raise UnknownMethodException('%s is not a valid method of interface %s' % (method_name, dbus_interface))
dbus.exceptions.UnknownMethodException: org.freedesktop.DBus.Error.UnknownMethod: Unknown method: getDeviceType is not a valid method of interface razer.device.misc
```

Daemon output:
```
2017-04-15 19:26:11 | razer                          | INFO     | Initialising Daemon (v1.1.9). Pid: 3609
2017-04-15 19:26:11 | razer.screensaver              | INFO     | Initialising DBus Screensaver Monitor
(...)
2017-04-15 19:26:11 | razer                          | DEBUG    | Loaded device specification: RazerMamba2012Wired --------------------- (1532:0024)
2017-04-15 19:26:11 | razer                          | DEBUG    | Loaded device specification: RazerMamba2012Wireless ------------------ (1532:0025)
(...)
2017-04-15 19:26:11 | razer                          | INFO     | Adding razer.devices.getDevices method to DBus
2017-04-15 19:26:11 | razer                          | INFO     | Adding razer.devices.enableTurnOffOnScreensaver method to DBus
2017-04-15 19:26:11 | razer                          | INFO     | Adding razer.devices.syncEffects method to DBus
2017-04-15 19:26:11 | razer                          | INFO     | Adding razer.devices.syncEffects method to DBus
2017-04-15 19:26:11 | razer                          | INFO     | Adding razer.devices.syncEffects method to DBus
2017-04-15 19:26:11 | razer                          | INFO     | Adding razer.daemon.version method to DBus
2017-04-15 19:26:11 | razer                          | INFO     | Adding razer.daemon.stop method to DBus
2017-04-15 19:26:11 | razer                          | INFO     | Serving DBus
2017-04-15 19:26:18 | razer                          | DEBUG    | Device event [add]: /devices/pci0000:00/0000:00:09.0/0000:02:00.0/usb2/2-4/2-4:1.0/0003:1532:0024.0004
2017-04-15 19:26:18 | razer                          | INFO     | Found valid device.0: 0003:1532:0024.0004
2017-04-15 19:26:18 | razer.device0                  | INFO     | Initialising device.0 RazerMamba2012Wired
2017-04-15 19:26:18 | razer.device0                  | DEBUG    | Adding getSerial method to DBus
2017-04-15 19:26:18 | razer.device0                  | DEBUG    | Adding razer.device.misc.suspendDevice method to DBus
2017-04-15 19:26:18 | razer.device0                  | DEBUG    | Adding razer.device.misc.getDeviceMode method to DBus
2017-04-15 19:26:18 | razer.device0                  | DEBUG    | Adding razer.device.misc.getRazerUrls method to DBus
2017-04-15 19:26:18 | razer.device0                  | DEBUG    | Adding razer.device.misc.setDeviceMode method to DBus
2017-04-15 19:26:18 | razer.device0                  | DEBUG    | Adding razer.device.misc.resumeDevice method to DBus
2017-04-15 19:26:18 | razer.device0                  | DEBUG    | Adding razer.device.misc.getVidPid method to DBus
2017-04-15 19:26:18 | razer.device0                  | DEBUG    | Adding razer.device.misc.version method to DBus
2017-04-15 19:26:18 | razer.device0                  | DEBUG    | Adding razer.device.misc.hasDedicatedMacroKeys method to DBus
2017-04-15 19:26:18 | razer.device0                  | DEBUG    | Adding razer.device.misc.getFirmware method to DBus
2017-04-15 19:26:18 | razer.device0                  | DEBUG    | Adding razer.device.misc.getMatrixDimensions method to DBus
2017-04-15 19:26:18 | razer.device0                  | DEBUG    | Adding razer.device.misc.hasMatrix method to DBus
2017-04-15 19:26:18 | razer.device0                  | DEBUG    | Adding razer.device.misc.getDeviceName method to DBus
2017-04-15 19:26:18 | razer.device0                  | DEBUG    | Adding razer.device.power.setLowBatteryThreshold method to DBus
2017-04-15 19:26:18 | razer.device0                  | DEBUG    | Adding razer.device.dpi.getDPI method to DBus
2017-04-15 19:26:18 | razer.device0                  | DEBUG    | Adding razer.device.dpi.setDPI method to DBus
2017-04-15 19:26:18 | razer.device0                  | DEBUG    | Adding razer.device.misc.getPollRate method to DBus
2017-04-15 19:26:18 | razer.device0                  | DEBUG    | Adding razer.device.misc.setPollRate method to DBus
2017-04-15 19:26:18 | razer.device0                  | DEBUG    | Adding razer.device.lighting.scroll.setScrollActive method to DBus
2017-04-15 19:26:18 | razer.device0                  | DEBUG    | Adding razer.device.lighting.scroll.getScrollActive method to DBus
2017-04-15 19:26:18 | razer.device0                  | DEBUG    | Adding razer.device.lighting.scroll.getScrollEffect method to DBus
2017-04-15 19:26:18 | razer.device0                  | DEBUG    | Adding razer.device.lighting.scroll.getScrollBrightness method to DBus
2017-04-15 19:26:18 | razer.device0                  | DEBUG    | Adding razer.device.lighting.scroll.setScrollBrightness method to DBus
2017-04-15 19:26:18 | razer.device0                  | DEBUG    | Adding razer.device.lighting.scroll.setScrollStatic method to DBus
2017-04-15 19:26:18 | razer.device0                  | DEBUG    | Adding razer.device.lighting.scroll.setScrollSpectrum method to DBus
2017-04-15 19:26:18 | razer                          | DEBUG    | Emitted Device Added Signal
2017-04-15 19:26:18 | razer                          | DEBUG    | Device event [add]: /devices/pci0000:00/0000:00:09.0/0000:02:00.0/usb2/2-4/2-4:1.1/0003:1532:0024.0005
```

**Comment by terricain** [comment]
```
2017-04-15 19:26:18 | razer.device0                  | DEBUG    | Adding razer.device.misc.getDeviceMode method to DBus
2017-04-15 19:26:18 | razer.device0                  | DEBUG    | Adding razer.device.misc.getRazerUrls method to DBus
2017-04-15 19:26:18 | razer.device0                  | DEBUG    | Adding razer.device.misc.setDeviceMode method to DBus
```
That looks like the issue. One sec

**Comment by lesderid** [comment]
Seems to be a missing comma here: https://github.com/terrycain/razer-drivers/blob/feature_mamba2012_0024_0025/daemon/razer_daemon/hardware/mouse.py#L596

**Comment by terricain** [comment]
Literally just noticed it :wink:

**Comment by terricain** [comment]
Whats it do after you fix the comma?

**Comment by lesderid** [comment]
Works great now. Controlling light colour (both spectrum and static), light brightness, and DPI all work as expected.

The only 'issue' is that the maximum DPI you can select is 16000, but the mouse only goes up to 6400 (setting it to anything higher than that doesn't do anything).

**Comment by terricain** [comment]
Ahh yeah, thats an issue me and the lads talked about earlier, Theres no way for us to know the max DPI unless someone goes on synapse and tells us. Will write that down and begin drafting something up to inform apps of DPI's that I'm aware of. But good :-) another device ready for release :smile:

**Comment by WhyNotHugo** [comment]
> the maximum DPI you can select is 16000, but the mouse only goes up to 6400

Are you sure? My mamba clearly says "16000DPI" at the bottom.

On the DPI side, maybe just hardcode the DPI each device supports (eg: map device_id to max_dpi)?

**Comment by terricain** [comment]
@hobarrera what mamba do you have? is it the same device id as this one

**Comment by WhyNotHugo** [comment]
No, no, that was my point: mine's the other one (`1532:0045`), and goes up to 16kdpi.

(sorry, I could have been clearer).

**Comment by terricain** [comment]
yeah older ones have less dpi i think

**Comment by lesderid** [comment]
Yeah, the 0024/0025 (i.e. this one, Mamba 2012) definitely only goes up to 6400.

**Comment by terricain** [comment]
Packages are building in launchpad, will be available soon
